tests/other/test_state_introspection.py

```diff
@@ -0,0 +1,94 @@
+
+ms_file = str(
+    Path(__file__).parent.parent.parent.joinpath("examples", "linux", "binaries", "multiple-styles")
+)
+
+
+class TestDaemonThread(unittest.TestCase):
+    def daemon(self, _thread):
+        while True:
+            self.fired = True
+            sleep(1)
+
+    def test_daemon(self):
+        self.fired = False
+        m = Manticore(ms_file, stdin_size=17)
+        m.register_daemon(self.daemon)
+        m.run()
+
+        self.assertTrue(self.fired)
+
+
+class TestIntrospect(unittest.TestCase):
+    def introspect_loop(self, thread):
+        while True:
+            self.history.append(thread.manticore.introspect())
+            sleep(0.5)
+
+    def test_introspect_daemon(self):
+        self.history = []
+        m = Manticore(ms_file, stdin_size=17)
+        m.register_daemon(self.introspect_loop)
+        m.run()
+
+        sleep(1)  # Leave time for the callback to fire after we've finished
+        self.assertGreater(len(self.history), 0)
+        progression = []
+        for hist in self.history:
+            hist = hist.values()
+            progression.append(
+                (
+                    sum(1 if (st.state_list == StateLists.ready) else 0 for st in hist),
+                    sum(1 if (st.state_list == StateLists.busy) else 0 for st in hist),
+                    sum(1 if (st.state_list == StateLists.terminated) else 0 for st in hist),
+                )
+            )
+        self.assertEqual(
+            progression[0][1] + progression[0][0], 1
+        )  # When fired for the first time, we have one busy state OR one ready state
+        self.assertGreater(progression[-1][2], 0)  # Once finished, we have some terminated states
+        self.assertEqual(progression[-1][0], 0)  # Once finished, we have no ready states
+        self.assertEqual(progression[-1][1], 0)  # Once finished, we have no ready states
+        # Once finished, we have only terminated states.
+        self.assertGreater(progression[-1][2], 0)
+
+        f = io.StringIO()
+        with contextlib.redirect_stdout(f):
+            m.pretty_print_states()
+        self.assertIn("Terminated States: {}".format(progression[-1][2]), f.getvalue())
+
+
+class MyIntrospector(IntrospectionAPIPlugin):
+    def on_execution_intermittent_callback(self, state, update_cb, *args, **kwargs):
+        super().on_execution_intermittent_callback(state, update_cb, *args, **kwargs)
+        with self.locked_context("manticore_state", dict) as context:
+            context[state.id].i_am_custom = True
+
+
+class TestCustomIntrospector(unittest.TestCase):
+    def introspect_loop(self, thread):
+        while True:
+            self.history.append(thread.manticore.introspect())
+            sleep(0.5)
+
+    def test_custom_introspector(self):
+        self.history = []
+        m = Manticore(ms_file, stdin_size=17)
+        m.set_instrospection_plugin(MyIntrospector)
+        m.register_daemon(self.introspect_loop)
+        m.run()
+
```

Comment:
... and instead keep the history in the "Introspector()"
```
class Introspector(Plugin):
     def get_history(self):
          with self.locked_context("blah", list) as l:
              return list(l)

     def at_introspection(state):
          with self.locked_context("blah", list) as l:
              l.append(state.get_status())

intro = Introspector("Specifc configuration")
m = Manticore(ms_file, stdin_size=17, plugins=(intro,))
with Thread(target=myfunc):
    m.run()
print (intro.get_history())
```

Reply:
We could, but is there a reason to keep the full history around long-term? It's nice to have for testing, but so far I've only thought of use cases for the introspection API where the most recent copy of the state descriptors matters. Would EVM benefit from this feature?

Reply:
I am clearly failing to understand something. 

I was thinking about a Plugin that implements whatever introspection and an event called "introspection".
Then the plugin will do whatever at that event (`at_instrospection()`), even keep or not the full history. And have a method to _print out_ the stats/results (`get_history()`).